src/csharp/Model.cs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.ML.OnnxRuntimeGenAI
+{
+    public enum DeviceType : long
+    {
+        Auto = 0,
+        CPU = 1,
+        CUDA = 2
+    }
+
+    public class Model : IDisposable
+    {
+        private IntPtr _modelHandle;
+
+        public Model(string modelPath, DeviceType deviceType)
+        {
+            Result.VerifySuccess(NativeMethods.OgaCreateModel(Utils.ToUtf8(modelPath), deviceType, out _modelHandle));
+        }
+
+        internal IntPtr Handle { get { return _modelHandle; } }
+
+        public int[][] Generate(GeneratorParams generatorParams)
```

Comment:
This creates a number of garbage objects. I think it would be much better to return ReadOnlySpan to the individual native buffers and let the user to decide what to do with them. 
To achive that introduce a Sequence class (and other classes as appriate) with the appropriate methods that return spans directly to native data.
This reduces the amount of garbage and eliminates copying.

Dealing with the multi-dim arrays is very inconvinient. Let the user build them if they want to.